File: python/fndsa/params.py
"""FN-DSA (FIPS 206 / FALCON) parameter sets."""

from dataclasses import dataclass

Q = 12289


@dataclass(frozen=True)
class Params:
    name: str
    n: int
    log_n: int
    padded: bool
    pk_size: int
    sk_size: int
    sig_size: int    # max for non-padded; exact for padded
    beta_sq: int
    fg_bits: int     # 6 for n=512, 5 for n=1024


FNDSA512 = Params("FN-DSA-512", 512, 9, False, 897, 1281, 666, 34034726, 6)
FNDSA1024 = Params("FN-DSA-1024", 1024, 10, False, 1793, 2305, 1280, 70265242, 5)
FNDSAPadded512 = Params("FN-DSA-PADDED-512", 512, 9, True, 897, 1281, 809, 34034726, 6)
FNDSAPadded1024 = Params("FN-DSA-PADDED-1024", 1024, 10, True, 1793, 2305, 1473, 70265242, 5)

File: js/fndsa/src/params.js
export const Q = 12289;

export class Params {
  constructor(name, n, logN, padded, pkSize, skSize, sigSize, betaSq, fgBits) {
    Object.assign(this, {name, n, logN, padded, pkSize, skSize, sigSize, betaSq, fgBits});
    Object.freeze(this);
  }
}

export const FNDSA512      = new Params("FN-DSA-512", 512, 9, false, 897, 1281, 666, 34034726, 6);
export const FNDSA1024     = new Params("FN-DSA-1024", 1024, 10, false, 1793, 2305, 1280, 70265242, 5);
export const FNDSAPadded512  = new Params("FN-DSA-PADDED-512", 512, 9, true, 897, 1281, 809, 34034726, 6);
export const FNDSAPadded1024 = new Params("FN-DSA-PADDED-1024", 1024, 10, true, 1793, 2305, 1473, 70265242, 5);
